File: 28数组中重复超过一半的数字/Solution.java
public class Solution {
    public int MoreThanHalfNum_Solution(int [] nums) {
        int majority = nums[0];
    for (int i = 1, cnt = 1; i < nums.length; i++) {
        cnt = nums[i] == majority ? cnt + 1 : cnt - 1;
        if (cnt == 0) {
            majority = nums[i];
            cnt = 1;
        }
    }
    int cnt = 0;
    for (int val : nums)
        if (val == majority)
            cnt++;
    return cnt > nums.length / 2 ? majority : 0;
    }
}
/* java解法是分为了两部分，首先利用逻辑通过一个遍历找到了出现最多的数字
然后判断这个数字出现的次数是不是超过有一半 */

File: 28数组中重复超过一半的数字/Solution.cpp
class Solution {
public:
    int MoreThanHalfNum_Solution(vector<int> numbers) {
        map<int,int>  num;
        int len =numbers.size();
        for (int i=0;i<len;i++){
            num[numbers[i]]++;
        }
        for (auto &v:num){
            if (v.second>len/2)
                return v.first;
        }
        return 0;
    }
};
/* 这个题用容器做就可以了，可以用map也可以用set，只要在
将所有的数字全部存入以后对数字进行统计就可以了，好在
map和set都存在统计的功能，map是通过键值对的更新统计
set可以通过其count函数进行统计。 */